pdns/recursordist/logging.cc

```diff
@@ -0,0 +1,179 @@
+/**
+ * This file is part of PowerDNS or dnsdist.
+ * Copyright -- PowerDNS.COM B.V. and its contributors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * In addition, for the avoidance of any doubt, permission is granted to
+ * link this program with OpenSSL and to (re)distribute the binaries
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "logging.hh"
+#include <string>
+#include "utility.hh"
+
+namespace Logging
+{
+
+std::shared_ptr<const Logger> Logger::getptr() const
+{
+  return shared_from_this();
+}
+
+bool Logger::enabled(Logr::Priority p) const
+{
+  return _level <= _verbosity || p != Logr::Absent;
+}
+
+void Logger::info(const std::string& msg) const
+{
+  logMessage(msg, Logr::Absent, boost::none);
+}
+
+void Logger::info(Logr::Priority p, const std::string& msg) const
+{
+  logMessage(msg, p, boost::none);
+}
+
+void Logger::logMessage(const std::string& msg, boost::optional<const std::string> err) const
+{
+  return logMessage(msg, Logr::Absent, err);
+}
+
+void Logger::logMessage(const std::string& msg, Logr::Priority p, boost::optional<const std::string> err) const
+{
+  if (!enabled(p)) {
+    return;
+  }
+  Entry entry;
+  entry.level = _level;
+  entry.d_priority = p;
+  Utility::gettimeofday(&entry.d_timestamp);
+  entry.name = _name;
+  entry.message = msg;
+  entry.error = err;
```

Comment:
Perhaps (but untested):
```suggestion
  entry.error = std::move(err);
```